bson/src/main/org/bson/codecs/pojo/PojoBuilderHelper.java

```diff
@@ -144,8 +152,61 @@
                 noArgsConstructor.setAccessible(true);
             }
         }
+        classModelBuilder.instanceCreatorFactory(new InstanceCreatorFactoryImpl<T>(
+                new ConstructorCreatorExecutable<>(clazz, noArgsConstructor)));
+    }
+
+    private static <T> void configureRecordClassModelBuilder(final ClassModelBuilder<T> classModelBuilder, final Class<T> clazz) {
+        classModelBuilder.isRecord(true);
+
+        Map<String, TypeParameterMap> propertyTypeParameterMap = new HashMap<>();
+        String declaringClassName = clazz.getSimpleName();
+        Map<String, PropertyMetadata<?>> propertyNameMap = new HashMap<>();
+
+        List<String> genericTypeNames = getGenericTypeNames(clazz);
+
+        List<Annotation> annotations = new ArrayList<>();
+        annotations.addAll(asList(clazz.getDeclaredAnnotations()));
+
+        RecordComponent[] recordComponents = clazz.getRecordComponents();
+        Class<?>[] types = new Class<?>[recordComponents.length];
+        for (int i = 0; i < recordComponents.length; i++) {
+            RecordComponent recordComponent = recordComponents[i];
+            String name = recordComponent.getName();
+            types[i] = recordComponent.getType();
+
+            Method getter;
+            try {
+                getter = clazz.getDeclaredMethod(name);
+            } catch (NoSuchMethodException e) {
+                throw new IllegalStateException("Record must have getter method with same name as field", e);
+            }
+            PropertyMetadata<?> metadata = getOrCreateMethodPropertyMetadata(name, declaringClassName, propertyNameMap,
+                    TypeData.newInstance(getter, true), propertyTypeParameterMap, null, genericTypeNames,
+                    getGenericType(getter, true));
+
+            metadata.setGetter(getter);
 
-        classModelBuilder.instanceCreatorFactory(new InstanceCreatorFactoryImpl<T>(new CreatorExecutable<T>(clazz, noArgsConstructor)));
+            for (Annotation a : recordComponent.getAnnotations()) {
```

Comment:
Because records are immutable, they won't have setters, so we can't really have the read/write breakdown we use for annotations in Pojos. Therefore I just add all the annotations to both the read and write annotations.

Also, by implementing it this way, I am requiring users to put the annotation on the RecordComponent (as opposed to a custom getter if they implement one). This seems reasonable to me, as overriding the getter (or writing a new one) should be pretty rare and I can't think of a reason a user would want to annotate the getter instead of the record component.